File: day15/soln.py
#!/usr/bin/python3
# Copyright (c) 2016 Bart Massey
# This program is licensed under the "MIT License".
# Please see the file COPYING in this distribution
# for license terms.

# Advent of Code Day 15. Broken prototype in Python.

from math import *

# https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm
def multinv(a, n):
    (t, new_t) = (0, 1)
    (r, new_r) = (n, a)
    while new_r != 0:
        q = r // new_r
        (t, new_t) = (new_t, t - q * new_t)
        (r, new_r) = (new_r, r - q * new_r)
    if r > 1:
        return None
    if t < 0:
        t += n
    return t

def lcm(x, y):
    return x * y // gcd(x, y)

test = None
if test == None:
    coeffs = [
        (7, 0),
        (13, 0),
        (3, 2),
        (5, 2),
        (17, 0),
        (19, 7) ]
elif test == 1:
    coeffs = [
        (5, 4),
        (2, 1) ]
elif test == 2:
    coeffs = [
        (5, 4),
        (2, 1),
        (3, 2) ]
elif test == 3:
    coeffs = [
        (5, 4),
        (2, 1),
        (17, 1),
        (3, 2),
        (11, 0) ]
else:
    assert(False)

c1 = coeffs[0][0]
q1 = coeffs[0][1]
t = None
for j in range(1, len(coeffs)):
    cj = coeffs[j][0]
    qj = coeffs[j][1]
    k1 = multinv(c1, cj) * (q1 - qj - 1)
    while k1 * c1 < q1 + j:
        k1 += cj
    t = k1 * c1 - q1 - j
    x = k1 * c1 + qj - q1 + 1
    assert(x % cj == 0)
    kj = x // cj
    q1 = c1 * cj - t
    c1 *= cj
print("math", t)

def search(t0):
    for j in range(len(coeffs)):
        cj = coeffs[j][0]
        qj = coeffs[j][1]
        t = t0 + 1 + j
        p = (t + qj) % cj
        if p != 0:
            return False
    return True

for ti in range(1000000):
    if search(ti):
        print("search", ti)
        break

File: day15/soln.rs
// Copyright © 2016 Bart Massey
// This program is licensed under the "MIT License".
// Please see the file COPYING in this distribution
// for license terms.

//! Advent of Code Day 15.

extern crate aoc;
extern crate captures_at;
use captures_at::CapturesAtExt;
extern crate regex;

/// Read the disc descriptions, returning
/// a vector of positions/start pairs.
fn read_discs() -> Vec<(isize, isize)> {
    // Set up the matching.
    let disc_pat = regex::Regex::new(
        r"^Disc #([0-9]+) has ([0-9]+) positions; at time=0, it is at position ([0-9]+).$",
    )
    .expect("main: could not compile disc pattern");

    // Set up the state.
    let mut discs = Vec::new();

    // Process the disc descriptions.
    for (ndiscs, target) in aoc::input_lines().enumerate() {
        // Parse the description.
        let parts = disc_pat.captures(&target).expect("malformed line");
        let disc_number: isize = parts
            .at(1)
            .expect("main: could not find disc number")
            .parse()
            .expect("main: could not parse disc number");
        let disc_positions = parts
            .at(2)
            .expect("main: could not find disc positions")
            .parse()
            .expect("main: could not parse disc positions");
        let disc_start = parts
            .at(3)
            .expect("main: could not find disc start")
            .parse()
            .expect("main: could not parse disc start");
        if disc_number != ndiscs as isize + 1 {
            panic!("unexpected disc number");
        };

        // Save the disc info.
        discs.push((disc_positions, disc_start));
    }
    discs
}

/// At the given start time `t0`, try to drop the ball
/// through all the discs and return success or failure.
fn search(discs: &[(isize, isize)], t0: isize) -> bool {
    for (j, &d) in discs.iter().enumerate() {
        let (cj, qj) = d;
        let t = t0 + 1 + j as isize;
        let p = (t + qj) % cj;
        if p != 0 {
            return false;
        };
    }
    true
}

/// Solve the problem.
pub fn main() {
    let discs = read_discs();
    for t in 0..std::isize::MAX {
        if search(&discs, t) {
            println!("{}", t);
            return;
        }
    }
    panic!("no solution found");
}
